File: Modulo12/Complexo/Complexo.cpp
#include "Complexo.h"

Complexo::Complexo(double real, double imaginaria) :
real(real), imaginaria(imaginaria) {
}

Complexo::~Complexo() {
}

void Complexo::imprime() {
    cout << *this << endl;
}

// operadores aritméticos binários

Complexo Complexo::operator+(const Complexo& obj) const {
    double r = this->real + obj.real;
    double i = this->imaginaria + obj.imaginaria;
    return Complexo(r, i);
}

Complexo Complexo::operator-(const Complexo& obj) const {
    return *this + -obj;
}

Complexo Complexo::operator*(const Complexo& obj) const {
    double a = this->real;
    double b = this->imaginaria;
    double c = obj.real;
    double d = obj.imaginaria;

    double r = a * c - b * d;
    double i = a * d + b * c;
    return Complexo(r, i);

    /* 
     * double r = this->real * obj.real - this->imaginaria * obj.imaginaria;
     * double i = this->real * obj.imaginaria + this->imaginaria * obj.real;
     * return Complexo(r,i);
     */

}

Complexo Complexo::operator/(const Complexo& obj) const {
    double a = this->real;
    double b = this->imaginaria;
    double c = obj.real;
    double d = obj.imaginaria;

    double r = (a * c + b * d) / (c * c + d * d);
    double i = (b * c - a * d) / (c * c + d * d);
    return Complexo(r, i);
}

// operador aritmético unário

Complexo Complexo::operator-() const {
    return Complexo(-this->real, -this->imaginaria);
}

// função friend (para impressão via cout)

ostream& operator<<(ostream& os, const Complexo& obj) {
    /* bool flag = obj.imaginaria >= 0;
    os << obj.real << (flag ? " + " : " - ");
    os << (flag ? obj.imaginaria : -obj.imaginaria) << "i"; */

    if (obj.imaginaria >= 0) {
        os << obj.real << " + " << obj.imaginaria << "i";
    } else {
        os << obj.real << " - " << -obj.imaginaria << "i";
    }
    return os;
}

File: Modulo09/Complexo-CSharp/Complexo.cs
namespace poo {
    
    public class Complexo {
        private double real;
        private double imaginaria;

        public Complexo(double real, double imaginaria) {
            this.real = real;
            this.imaginaria = imaginaria;
        }

        public override string ToString() {
            bool flag = this.imaginaria >= 0;
            string op = (flag ? " + " : " - ");
            double img = (flag ? this.imaginaria : -this.imaginaria);

            return string.Format("{0}{1}{2}i", this.real, op, img);
        }

        // operadores aritméticos binários

        public static Complexo operator +(Complexo c1, Complexo c2) {
            double r = c1.real + c2.real;
            double i = c1.imaginaria + c2.imaginaria;
            return new Complexo(r, i);
        }

        public static Complexo operator -(Complexo c1, Complexo c2) {
            return c1 + -c2;
        }

        public static Complexo operator *(Complexo c1, Complexo c2) {
            double a = c1.real;
            double b = c1.imaginaria;
            double c = c2.real;
            double d = c2.imaginaria;

            double r = a * c - b * d;
            double i = a * d + b * c;

            return new Complexo(r, i);
        }

        public static Complexo operator /(Complexo c1, Complexo c2) {
            double a = c1.real;
            double b = c1.imaginaria;
            double c = c2.real;
            double d = c2.imaginaria;

            double r = (a * c + b * d) / (c * c + d * d);
            double i = (b * c - a * d) / (c * c + d * d);

            return new Complexo(r, i);
        }

        // operador aritméticos unários

        public static Complexo operator -(Complexo c1) {
            return new Complexo(-c1.real, -c1.imaginaria);
        }
    }

}